### PR TITLE
[SPARK-26836][SQL] Supporting Avro schema evolution for partitioned Hive tables with "avro.schema.literal"

### PR DESCRIPTION
**Pull request by attilapiros**
### What changes were proposed in this pull request?

Before this PR for a partitioned Avro Hive table when the SerDe is configured to read the partition data
the table level properties were overwritten by the partition level properties.

This PR changes this ordering by giving table level properties higher precedence  thus when a new evolved schema 
is set for the table this new schema will be used to read the partition data and not the original schema which was used for writing the data.

This new behavior is consistent with Apache Hive. 
See the example used in the unit test `SPARK-26836: support Avro schema evolution`, in Hive this results in: 

```
0: jdbc:hive2://<IP>:10000> select * from t;
INFO  : Compiling command(queryId=hive_20210111141102_7a6349d0-f9ed-4aad-ac07-b94b44de2394): select * from t
INFO  : Semantic Analysis Completed
INFO  : Returning Hive schema: Schema(fieldSchemas:[FieldSchema(name:t.col1, type:string, comment:null), FieldSchema(name:t.col2, type:string, comment:null), FieldSchema(name:t.ds, type:string, comment:null)], properties:null)
INFO  : Completed compiling command(queryId=hive_20210111141102_7a6349d0-f9ed-4aad-ac07-b94b44de2394); Time taken: 0.098 seconds
INFO  : Executing command(queryId=hive_20210111141102_7a6349d0-f9ed-4aad-ac07-b94b44de2394): select * from t
INFO  : Completed executing command(queryId=hive_20210111141102_7a6349d0-f9ed-4aad-ac07-b94b44de2394); Time taken: 0.013 seconds
INFO  : OK
+---------------+-------------+-------------+
|    t.col1     |   t.col2    |    t.ds     |
+---------------+-------------+-------------+
| col1_default  | col2_value  | 1981-01-07  |
| col1_value    | col2_value  | 1983-04-27  |
+---------------+-------------+-------------+
2 rows selected (0.159 seconds)
```

### Why are the changes needed?

Without this change the old schema would be used. This can use a correctness issue when the new schema introduces 
a new field with a default value (following the rules of schema evolution) before an existing field. 
In this case the rows coming from the partition where the old schema was used will **contain values in wrong column positions**.

For example check the attached unit test `SPARK-26836: support Avro schema evolution` 

Without this fix the result of the select on the table would be:

```
+----------+----------+----------+
|      col1|      col2|        ds|
+----------+----------+----------+
|col2_value|      null|1981-01-07|
|col1_value|col2_value|1983-04-27|
+----------+----------+----------+

```

With this fix:

```
+------------+----------+----------+
|        col1|      col2|        ds|
+------------+----------+----------+
|col1_default|col2_value|1981-01-07|
|  col1_value|col2_value|1983-04-27|
+------------+----------+----------+
```

### Does this PR introduce _any_ user-facing change?

Just fixes the value errors. 
When a new column is introduced even to the last position then instead of 'null' the given default will be used. 

### How was this patch tested?

This was tested with the unit tested included to the PR. 
And manually on Apache Spark / Hive.

### CODE REVIEW

**Comment by attilapiros** [comment]
jenkins retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 1]
```diff
@@ -1883,6 +1883,60 @@ class HiveDDLSuite
...
+      sql(
+        s"""
+           |CREATE TABLE t PARTITIONED BY (ds string)
+           |ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.avro.AvroSerDe'
+           |WITH SERDEPROPERTIES ('avro.schema.literal'='$originalSchema')
+           |STORED AS
+           |INPUTFORMAT 'org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat'
+           |OUTPUTFORMAT 'org.apache.hadoop.hive.ql.io.avro.AvroContainerOutputFormat'
+        """.stripMargin)
+      sql(s"INSERT INTO t partition (ds='1981-01-07') VALUES ('col2_value')")
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1911]: nit. `s"` -> `"`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 2]
```diff
@@ -1883,6 +1883,60 @@ class HiveDDLSuite
...
+          |    },
+          |    {
+          |      "name": "col2",
+          |      "type": "string"
+          |    }
+          |  ]
+          |}
+        """.stripMargin
+      sql(s"""ALTER TABLE t SET SERDEPROPERTIES ('avro.schema.literal'='$evolvedSchema')""")
+      sql(s"INSERT INTO t partition (ds='1983-04-27') VALUES ('col1_value', 'col2_value')")
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1932]: nit. `s"` -> `"`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 3]
```diff
@@ -1883,6 +1883,60 @@ class HiveDDLSuite
...
+      withSQLConf("spark.sql.hive.avroSchemaEvolution.enabled" -> "true") {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1933]: `HIVE_AVRO_SCHEMA_EVOLUTION_ENABLED.key`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 4]
```diff
@@ -1883,6 +1883,60 @@ class HiveDDLSuite
...
+    withTable("t") {
+      val originalSchema =
+        """
+          |{
+          |  "namespace": "test",
+          |  "name": "some_schema",
+          |  "type": "record",
+          |  "fields": [
+          |    {
+          |      "name": "col2",
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1896]: Just a question because it's a little counter-intuitive. Is there a reason why we start with `col2` and add `col1` later? Usually, I thought we start with `col1` and add `col2` later.

**Comment by HyukjinKwon** [comment]
cc @gengliangwang FYI

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 4 above)
Comment (attilapiros) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1896]: Yes, regarding schema evolution rules you can add a field to an arbitrary position, see the first example here:
https://docs.oracle.com/database/nosql-12.1.3.0/GettingStartedGuide/schemaevolution.html
 
My intention with this example and field naming was to illustrate the worst case (the column mismatch error) and emphasize its root cause. And even by adding new fields at the end one just decrease the scope of the problem as there will be still wrong values (null) for the new fields. 

Moreover when an existing field is removed the column mismatch usually cannot be avoided, example:

```
sql("""
  CREATE TABLE t PARTITIONED BY (ds string)
  ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.avro.AvroSerDe'
  WITH SERDEPROPERTIES ('avro.schema.literal'='
  {
    "namespace": "test",
    "name": "some_schema",
    "type": "record",
    "fields": [
      {
        "name": "col1",
        "type": "string",
        "default": "col1_default"
      },
      {
        "name": "col2",
        "type": "string"
      }
    ]
  }')
  STORED AS
  INPUTFORMAT 'org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat'
  OUTPUTFORMAT 'org.apache.hadoop.hive.ql.io.avro.AvroContainerOutputFormat'
""")

sql("""
  INSERT INTO t partition (ds='1981-01-07') VALUES ('col1_value', 'col2_value')
""")

sql("""
  ALTER TABLE t SET SERDEPROPERTIES ('avro.schema.literal'='
  {
    "namespace": "test",
    "name": "some_schema",
    "type": "record",
    "fields": [
      {
        "name": "col2",
        "type": "string"
      }
    ]
  }')
""")

sql("""
  INSERT INTO t partition (ds='1983-04-27') VALUES ('col2_value')
""")

sql("""
  select * from t
""").show()
```

Without this PR:
```
+------------+----------+
|        col2|        ds|
+------------+----------+
|col1_default|1981-01-07|
|  col2_value|1983-04-27|
+------------+----------+
```

With the fix:
```
+----------+----------+
|      col2|        ds|
+----------+----------+
|col2_value|1981-01-07|
|col2_value|1983-04-27|
+----------+----------+
```

**Comment by attilapiros** [comment]
@dongjoon-hyun I have addressed all of your comments, is there anything else I can help you with?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 5]
```diff
@@ -1883,6 +1883,60 @@ class HiveDDLSuite
...
+          |    }
+          |  ]
+          |}
+        """.stripMargin
+      sql(s"""ALTER TABLE t SET SERDEPROPERTIES ('avro.schema.literal'='$evolvedSchema')""")
+      sql("INSERT INTO t partition (ds='1983-04-27') VALUES ('col1_value', 'col2_value')")
+      withSQLConf(SQLConf.HIVE_AVRO_SCHEMA_EVOLUTION_ENABLED.key -> "true") {
+        checkAnswer(spark.table("t"), Row("col1_default", "col2_value", "1981-01-07")
+          :: Row("col1_value", "col2_value", "1983-04-27") :: Nil)
+      }
```
Comment (xkrogen) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1936]: Maybe also `checkAnswer` without the new schema evolution config enabled?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 6]
```diff
@@ -248,11 +250,17 @@ class HadoopTableReader(
...
+        // "spark.sql.hive.avroSchemaEvolution.enabled" is set because in that case the properties
+        // given at table level will be used (for details please check SPARK-26836).
+        // For example, a partition may have a different SerDe as the one defined in table
+        // properties.
         val props = new Properties(tableProperties)
-        partProps.asScala.foreach {
-          case (key, value) => props.setProperty(key, value)
+        partProps.asScala.foreach { case (key, value) =>
+          if (!avroSchemaEvolutionEnabled ||
+            !avroTablePropertyKeys.contains(key) || !tableProperties.containsKey(key)) {
```
Comment (xkrogen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:261]: Can we rephrase this as `!(avroSchemaEvolutionEnabled && avroTablePropertyKeys.contains(key) && tableProperties.containsKey(key))` ? I think the disjunction of negations is more confusing.

You can also consider something like:
```
partProps.asScala
  .filterNot { case (key, _) => avroSchemaEvolutionEnabled && avroTablePropertyKeys.contains(key) && ... }
  .foreach { case (key, value) => props.setProperty(key, value)
```
but I'm not sure if that is more readable or not.

**Review by xkrogen — CHANGES_REQUESTED** [review]
Good catch @attilapiros, it's great to see this!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 5 above)
Comment (attilapiros) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1936]: I have my doubts regarding this: I mean to validate a bad behavior where one value is given in the wrong column. 
Do we want to freeze the current bad behavior? and why would be that specific wrong result better than all the others?

Comment (xkrogen) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1936]: We have the config default to `false` to ensure we preserve the old behavior for compatibility reasons, so it seems to me like we should have a test to verify that we have actually preserved the old behavior. A bug present for long enough eventually starts to become a feature 🤷🏼‍♂️

Basically my take is: if we don't care about compatibility then we shouldn't even have a config, if we do care about compatibility then we should validate it with a test.

But your point is valid so I am open to differing opinions.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 6 above)
Comment (attilapiros) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: I like the more functional solution with the `filterNot`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 5 above)
Comment (attilapiros) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1936]: I mean there are different use cases depending on how the schema is changed. 

Here we have an example where the value in wrong column makes the old behavior practically unusable. 

But for cases when just the default value is changed that sounds at least acceptable: 
all the data is read back with the value which was used during the write (including the missing values).
What about adding such a case for the `SQLConf.HIVE_AVRO_SCHEMA_EVOLUTION_ENABLED.key -> "false"`?

Comment (xkrogen) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1936]: Yes, that sounds perfect. I think this is my fault for not reading the test case closely enough and thus making a bad suggestion, sorry about that. Thanks for figuring out what I really wanted :)

Comment (attilapiros) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1936]: During the development of our new test I realized to use default value of the schema is only possible via adding a new column. As NULL is different from the missing value thus will not be mapped to default but to the "null" value in the file.
And INSERT always requires to have a value specified  for each of the column.

Based on these new findings I am deplementing `spark.sql.hive.avroSchemaEvolution.enabled` and always use the table schema when available.

**Comment by attilapiros** [comment]
cc @cloud-fan

**Comment by attilapiros** [comment]
@dongjoon-hyun may I ask for another review from you?

**Comment by dongjoon-hyun** [comment]
Sure, I'll review right now, @attilapiros .

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 7]
```diff
@@ -388,6 +394,9 @@ private[hive] object HiveTableUtil {
 private[hive] object DeserializerLock
 
 private[hive] object HadoopTableReader extends HiveInspectors with Logging {
+
+  val avroTableProperties = AvroTableProperties.values().map(_.getPropName()).toSet
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:398]: SPARK-26836 is filed for 2.3.1 ~ 3.2.0 and this patch depends on `org.apache.hadoop.hive.serde2.avro.AvroSerdeUtils.AvroTableProperties`. I'm wondering if this patch is applicable with old Spark branches with old Hive libraries. For example,
- Apache Spark master : 2.3.8
- Apache Spark branch-3.1: 2.3.7
- Apache Spark branch-3.0: 2.3.7 and 1.2.1.spark2
- Apache Spark branch-2.4: 1.2.1.spark2

What is the minimum version of Hive library to use this? Is there a known bug for this `AvroTableProperties` classes?

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:398]: According to Hive communtiy, Apache Spark cannot access the following two table properties because it's in Apache Hive 4 via HIVE-21218 . I'm wondering if this can cause some issues in the future.
```
AVRO_SERDE_TYPE("avro.serde.type"),
AVRO_SERDE_SKIP_BYTES("avro.serde.skip.bytes"),
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
According to the PR scope, it seems that we only need `avro.schema.literal`. Did I understand correctly? In that case, let's not depend on `org.apache.hadoop.hive.serde2.avro.AvroSerdeUtils.AvroTableProperties`. I'd recommend to use `avro.schema.literal` directly and minimally in this PR.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 7 above)
Comment (attilapiros) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:398]: The `AvroTableProperties` enum is used for avoiding the hardcoding of those property names which related to Avro and thus where the table properties has higher priority when a value for that property is chosen.

So this PR does NOT set any of those properties if the user hasn't specified one by using the `SERDEPROPERTIES` on the table or on the partition here we just change which one is wining before passing the values further to the SerDe initialization.

The `avro.schema.literal` on its own is not enough for example if the user used different `avro.schema.url` on the table than on the partition we will have the same issue. This can be true for all the properties it is good if they are coming from the consistent source: either from the table or from the partition. The advantage of using `AvroTableProperties` that we should not investigate all the consistency between these properties but just take them all from the same source and when a new one is introduced in the future which matters for the serde we will be still safe and this part does not generate a new bug silently (as we cannot have test for not existing future serde properties).

What do you think?

Comment (attilapiros) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:398]: @dongjoon-hyun I can offer a compromised solution where `AvroTableProperties` is not used at all. Instead of that the Avro property names will be detected with a `startsWith("avro.")` check.

 Is it ok with you?

Comment (xkrogen) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:398]: It seems that the examples of `avro.serde.type` and `avro.serde.skip.bytes` provide good reason why we shouldn't pick up `AvroTableProperties.*` or `startsWith("avro.")`. These properties indicate something about the specific files, e.g. it's possible you could have some partitions serialized with the Confluence Avro serializer (thus requiring the `avro.serde` properties) and some partitions serialized with normal Avro behavior. Schemas seem to be a special case.

Looking at the full list of properties in Hive 2.3:
```
    SCHEMA_LITERAL("avro.schema.literal"),
    SCHEMA_URL("avro.schema.url"),
    SCHEMA_NAMESPACE("avro.schema.namespace"),
    SCHEMA_NAME("avro.schema.name"),
    SCHEMA_DOC("avro.schema.doc"),
    AVRO_SERDE_SCHEMA("avro.serde.schema"),
    SCHEMA_RETRIEVER("avro.schema.retriever");
```
It seems that the first 5 we would definitely want at the table level, and probably the last one (?). I don't understand `avro.serde.schema` and how it relates to the others. But it might be reasonable to say that we should take all `startsWith("avro.schema.")`


Separately from the question of _which_ properties to pick up, there is the question of whether to hard-code the strings or use the enum values. @dongjoon-hyun can you explain more fully why you don't want to depend on `AvroTableProperties`? We already have another reference to it in the Spark codebase [within HiveShim](https://github.com/apache/spark/blob/6a194f197819e2970f619ed92321e0a0b716bf37/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala#L92). I don't understand the argument that some older branches may not have this class available -- if an older branch has a Hive dependency that is too old and we need to cherry pick this change there, then we can hard-code it in the cherry pick, but I don't see a reason why we can't use the enum in newer versions.

Comment (attilapiros) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:398]: @xkrogen thanks for the correct pattern: `startsWith("avro.schema.")`!

I will use that one.

Comment (xkrogen) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:398]: @attilapiros would you mind looking into what `avro.serde.schema` is used for? Understanding that better would provide more confidence (for me, at least) that `avro.schema` is the correct prefix.

**Comment by dongjoon-hyun** [comment]
@xkrogen . That another instance explicitly uses `AvroSerdeUtils.AvroTableProperties.SCHEMA_LITERAL.getPropName()`, not other property. My suggestion, `it seems that we only need avro.schema.literal`, was the same.

>  @dongjoon-hyun can you explain more fully why you don't want to depend on AvroTableProperties? We already have another reference to it in the Spark codebase within HiveShim.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 7 above)
Comment (attilapiros) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Sorry I read this message after my git push, but sure I will check that.

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @attilapiros . But, I'm still not sure about the other properties in those namespace.
For `avro.schema.literal`, your test case provides the value. For the others, I'm negative if there is no useful test case in the Apache Spark community.

So, if you don't mind, may I ask again you narrow down this PR to `avro.schema.literal` first. For the other new configurations, please proceed separately if you want.

**Comment by attilapiros** [comment]
@dongjoon-hyun this sound good to me. 
Here I will use explicitly `AvroSerdeUtils.AvroTableProperties.SCHEMA_LITERAL.getPropName()` and for the rest I will open separate PRs with new tests proving they are useful.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 7 above)
Comment (xkrogen) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Based on new discussion, sounds like we can leave that as a follow-on for later.

**Comment by xkrogen** [comment]
Thanks for the clarification @dongjoon-hyun ! I understand your concern now.

New plan sounds good to me as well.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 7 above)
Comment (attilapiros) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Yes, but thanks for all your inputs.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 8]
```diff
@@ -388,6 +393,7 @@ private[hive] object HiveTableUtil {
 private[hive] object DeserializerLock
 
 private[hive] object HadoopTableReader extends HiveInspectors with Logging {
+
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:396]: Could you remove this to reduce the diff?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 9]
```diff
@@ -239,7 +240,6 @@ class HadoopTableReader(
       fillPartitionKeys(partValues, mutableRow)
 
       val tableProperties = tableDesc.getProperties
-
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:242]: Let's not remove this line~

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 10]
```diff
@@ -248,11 +248,16 @@ class HadoopTableReader(
...
+        // properties take a higher priority here except for the Avro table properties
+        // to support schema evolution: in that case the properties given at table level will
+        // be used (for details please check SPARK-26836).
+        // For example, a partition may have a different SerDe as the one defined in table
+        // properties.
         val props = new Properties(tableProperties)
-        partProps.asScala.foreach {
-          case (key, value) => props.setProperty(key, value)
+        partProps.asScala.filterNot { case (k, _) =>
+          k == AvroTableProperties.SCHEMA_LITERAL.getPropName() && tableProperties.containsKey(k)
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:259]: `&& tableProperties.containsKey(k)` looks risky to me and beyond this PR's test coverage. According to the test case, it looks like `k == AvroTableProperties.SCHEMA_LITERAL.getPropName()` is enough to pass the test case. Did I understand correctly?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 11]
```diff
@@ -248,11 +249,16 @@ class HadoopTableReader(
...
+        }.foreach { case (key, value) =>
+          props.setProperty(key, value)
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:261]: Let's revert to the original form like the following. It will make your PR content more clear.
```scala
-        }.foreach { case (key, value) =>
-          props.setProperty(key, value)
+        }.foreach {
+          case (key, value) => props.setProperty(key, value)
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 12]
```diff
@@ -1883,6 +1883,58 @@ class HiveDDLSuite
     }
   }
 
+  test("SPARK-26836: support Avro schema evolution") {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1886]: Could you add an opposite test case which column removal schema evolution, please?

Comment (dongjoon-hyun) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1886]: `Avro` is known to support (1), (2), (3) from the following four.
```
 *   1. Add a column
 *   2. Hide a column
 *   3. Change a column position
 *   4. Change a column type (Upcast)
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 11 above)
Comment (attilapiros) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:261]: Even when 2 lines above the other style looks good to avoid the 100 character limit for lines?

So do you prefer this?

```
          partProps.asScala.filterNot { case (k, _) =>
            k == AvroTableProperties.SCHEMA_LITERAL.getPropName() && tableProperties.containsKey(k)
          }.foreach {
            case (key, value) => props.setProperty(key, value)
          }
```

**Review by dongjoon-hyun — COMMENTED** [review]
At data source layer, we have the following schema evolution test coverage. I guess we need at least (1) `add column` and (2) `hide column` test case in this PR. (1) is already included in this PR. So, we need (2).
- https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala
```
 *   | File Format  | Coverage     | Note                                                   |
 *   | ------------ | ------------ | ------------------------------------------------------ |
 *   | TEXT         | N/A          | Schema consists of a single string column.             |
 *   | CSV          | 1, 2, 4      |                                                        |
 *   | JSON         | 1, 2, 3, 4   |                                                        |
 *   | ORC          | 1, 2, 3, 4   | Native vectorized ORC reader has the widest coverage.  |
 *   | PARQUET      | 1, 2, 3      |                                                        |
 *   | AVRO         | 1, 2, 3      |                                                        |
```

If you can add (3) `Change a column position`, it's the best.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 11 above)
Comment (attilapiros) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:261]: I have chosen the consistent over keeping the old style. Or should I modify the `filterNot` body too?

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:261]: Yes, that one.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 12 above)
Comment (attilapiros) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1886]: It's 22:53 here so the rest of the requirements remains for the future.

Comment (dongjoon-hyun) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala:1886]: Sure, take your time~

Comment (attilapiros) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: I have added a new unit test: `SPARK-26836: support Avro schema evolution (remove column)`

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @attilapiros and all.

Merged to master.

cc @gengliangwang , too.
